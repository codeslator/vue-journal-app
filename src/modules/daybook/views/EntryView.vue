<template>
  <template v-if="entry">
    <div class="card entry-card">
      <div class="row g-0">
        <div class="col-4">
          <img
            v-if="entry.picture && !localImage"
            :src="entry.picture"
            alt=""
            class="img-fluid rounded-start"
          />
          <img
            v-if="localImage"
            :src="localImage"
            alt=""
            class="img-fluid rounded-start"
          />
          <input
            type="file"
            @change="selectImage"
            name="picture"
            id="picture"
            ref="imageSelector"
            v-show="false"
            accept="image/png, image/jpg"
          />
        </div>
        <div :class="entry.picture || localImage ? 'col-8' : 'col-12'">
          <div
            class="
              card-header
              d-flex
              align-items-center
              justify-content-between
            "
          >
            <div class="entry-title">
              <h5 class="card-title text-success mr-2">
                {{ month }} {{ day }}
              </h5>
              <p class="card-subtitle text-muted">{{ yearDay }}</p>
            </div>
            <div class="entry-actions">
              <button class="btn btn-info" @click="selectingImage">
                Upload Picture
                <i class="fa fa-upload"></i>
              </button>
              <button
                class="btn btn-danger"
                v-if="entry.id"
                @click="removeEntry"
              >
                Delete
                <i class="fa fa-trash-alt"></i>
              </button>
            </div>
          </div>
          <div class="card-body">
            <div class="row mt-2">
              <div class="col">
                <textarea
                  class="form-control"
                  placeholder="What happened today?"
                  name=""
                  id=""
                  v-model="entry.text"
                ></textarea>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </template>
  <fab @on:click="storeEntry" :icon="'fa-save'" />
</template>

<script>
import { defineAsyncComponent } from "vue";
import { mapGetters, mapActions } from "vuex";
import Swal from "sweetalert2";
import getDayMonthYear from "../helpers/getDayMonthYear";
import uploadImage from "../helpers/uploadImage";

export default {
  components: {
    Fab: defineAsyncComponent(() =>
      import(
        /* webpackChunkName: "DaybookFab" */ "@/modules/daybook/components/Fab"
      )
    ),
  },
  props: {
    id: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      entry: null,
      localImage: null,
      file: null,
    };
  },
  computed: {
    ...mapGetters("journal", {
      getEntry: "getEntry",
    }),
    day() {
      const { day } = getDayMonthYear(this.entry.date);
      return day;
    },
    month() {
      const { month } = getDayMonthYear(this.entry.date);
      return month;
    },
    yearDay() {
      const { yearDay } = getDayMonthYear(this.entry.date);
      return yearDay;
    },
  },
  methods: {
    ...mapActions("journal", ["updateEntry", "saveEntry", "deleteEntry"]),
    loadEntry() {
      let entry;
      if (this.id === "new") {
        entry = {
          text: "",
          date: new Date().getTime(),
          // picture: null,
        };
      } else {
        entry = this.getEntry(this.id);
        if (!entry) return this.$router.push({ name: "no-entry" });
      }
      this.entry = entry;
    },
    async storeEntry() {
      new Swal({
        title: "Please wait...",
        allowOutsideClick: false,
      });
      Swal.showLoading();
      const picture = await uploadImage(this.file);
      this.entry.picture = picture;
      if (this.entry.id) {
        await this.updateEntry(this.entry);
        Swal.fire({
          title: "Updated",
          text: "Entry updated successfully.",
          icon: "success",
        });
      } else {
        const id = await this.saveEntry(this.entry);
        this.$router.push({ name: "entry", params: { id: id } });
        Swal.fire({
          title: "Saved",
          text: "Entry saved successfully.",
          icon: "success",
        });
      }
      this.file = null;
    },
    async removeEntry() {
      const { isConfirmed } = await Swal.fire({
        title: "Are you sure?",
        text: "Once you did it you can't recover data",
        showDenyButton: true,
        confirmButtonText: "I'm sure",
      });
      if (isConfirmed) {
        new Swal({
          title: "Please wait...",
          allowOutsideClick: false,
        });
        Swal.showLoading();
        await this.deleteEntry(this.entry.id);
        this.$router.push({ name: "no-entry" });
        Swal.fire({
          title: "Deleted",
          text: "Entry deleted successfully.",
          icon: "success",
        });
      }
    },
    selectImage(event) {
      const file = event.target.files[0];
      if (!file) {
        this.localImage = null;
        this.file = null;
        return;
      }
      this.file = file;
      const fileReader = new FileReader();
      fileReader.onload = () => (this.localImage = fileReader.result);
      fileReader.readAsDataURL(file);
    },
    selectingImage() {
      this.$refs.imageSelector.click();
    },
  },
  created() {
    this.loadEntry();
  },
  watch: {
    id() {
      this.loadEntry();
    },
  },
};
</script>

<style lang="scss" scoped>
.entry-picture {
  width: 200px;
  position: fixed;
  right: 10px;
}
</style>