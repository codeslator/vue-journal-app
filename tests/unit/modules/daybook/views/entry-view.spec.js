import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import Swal from 'sweetalert2';
import { daybookState } from '../../../mocks/daybook-state';
import journalStore from '@/modules/daybook/store/';
import EntryView from '@/modules/daybook/views/EntryView';

const createVuexStore = (initialState) => createStore({
  modules: {
    journal: {
      ...journalStore,
      state: { ...initialState }
    },
  }
});

jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn()
})); // Crea un mock de la instancia de la libreria y retorna l;as funciones y datos del objeto

describe('Entry View', () => {
  const store = createVuexStore(daybookState);
  store.dispatch = jest.fn();
  let wrapper;

  const mockRouter = {
    push: jest.fn(),
  };


  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryView, {
      props: {
        id: daybookState.entries[0].id
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store]
      }
    });
  })

  test('Should to redirect if id doesn\'t exists', () => {
    const wrapper = shallowMount(EntryView, {
      props: {
        id: 'id not found'
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store]
      },
    });

    expect(mockRouter.push).toHaveBeenCalledWith({ name: "no-entry" });
  });

  test('Should to show the entry correctly and match with last snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(mockRouter.push).not.toHaveBeenCalled();
  });


  test('Should to delete the entry and the exit', (done) => {
    Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }));
    const button =  wrapper.find('.btn-danger');
    button.trigger('click');
    expect(Swal.fire).toHaveBeenCalledWith({
      title: "Are you sure?",
      text: "Once you did it you can't recover data",
      showDenyButton: true,
      confirmButtonText: "I'm sure",
    });
    setTimeout(() => {
      expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-MojfF3krYcTW3t1GL6Z')
      expect(mockRouter.push).toHaveBeenCalled();
      done();
    }, 1);
  })
});
