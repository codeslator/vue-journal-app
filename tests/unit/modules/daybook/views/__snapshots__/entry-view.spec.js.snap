// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry View Should to show the entry correctly and match with last snapshot 1`] = `
<div class="card entry-card">
  <div class="row g-0">
    <div class="col-4"><img src="https://res.cloudinary.com/codeslator/image/upload/v1637182405/journal-app/qzqn1sci7pby0yvcft2s.jpg" alt="" class="img-fluid rounded-start">
      <!--v-if--><input type="file" name="picture" id="picture" accept="image/png, image/jpg" style="display: none;">
    </div>
    <div class="col-8">
      <div class="card-header d-flex align-items-center justify-content-between">
        <div class="entry-title">
          <h5 class="card-title text-success mr-2">Noviembre 3</h5>
          <p class="card-subtitle text-muted">2021, Mi√©rcoles</p>
        </div>
        <div class="entry-actions"><button class="btn btn-info"> Upload Picture <i class="fa fa-upload"></i></button><button class="btn btn-danger"> Delete <i class="fa fa-trash-alt"></i></button></div>
      </div>
      <div class="card-body">
        <div class="row mt-2">
          <div class="col"><textarea class="form-control" placeholder="What happened today?" name="" id=""></textarea></div>
        </div>
      </div>
    </div>
  </div>
</div>
<fab-stub icon="fa-save"></fab-stub>
`;
